Avoid unnecessary temporary `data` tensor in Saved Variable

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62410
* #61928
* __->__ #61927

This is a refactor of `SavedVariable.cpp` to prevent ever defining the `data_` tensor if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only needed if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

Varal7 committed:
Avoid unnecessary copying data in Saved Variable

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

soulitzer reviewed (commented):
The refactor looks pretty good. Could you add in the PR description which copy is no longer needed?

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data();
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data();
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data();
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data();
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data();
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only copied if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only copied if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only copied if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

zou3519 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -90,6 +98,9 @@ void SavedVariable::save_common_metadata(const Variable& data) {
   }
 }
 
+std::unique_ptr<SavedVariableHooks> SavedVariable::get_default_hooks_() {
+  return Engine::get_default_engine().get_default_saved_variable_hooks();
+}
```
I'm not familiar with this piece of code so going to ask a lot of noob questions here...
- Why does this function end in `_`?
- Why does this return a unique_ptr? If we call `get_default_hooks_()` twice, then it returns two different sets of default hooks?
- Should this be a const method? Does this modify the state of SavedVariable?

Varal7 replied:
1.  No need for `_`, I will remove them
2. I'm using unique_ptr because it makes it easier to guarantee that the underlying `PyObject* pack_hook` and `unpack_hook` are correctly increfed and decrefed. This does create a new `SavedVariableHooks` object with its own pair of hooks for each `SavedVariable`, but we need that in order for the hooks to survive when default hooks are unset but the `SavedVariable` is still alive. 
As for the current code, it is true that we might be missing out on a optimization by  doing `if (get_default_hooks()) {` instead of something like `if (has_default_hooks()) {`. Do you think it's worth the code duplication?
3. This could indeed be a const method

zou3519 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -43,6 +43,18 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     is_output_ = is_output;
+    is_inplace_on_view_ = is_inplace_on_view;
+
+    if(is_inplace_on_view) {
+      TORCH_INTERNAL_ASSERT(!is_leaf_ && is_output);
+      weak_grad_fn_ = variable.grad_fn();
+    }
```
The previous comment was `// From now on, we can assume the variable is not a leaf and is an output.`. Why was it fine to move this code up from where it used to be?

Varal7 replied:
The short answer is:
A saved variable can be an `is_inplace_on_view` only if it `is_output` and not `is_leaf`.
If `is_inplace_on_view` is true, we will never run into the code about `saved_original`, so it's fine.
If `is_inplace_on_view` is ƒalse, that snippet of code was and still is a no op.

zou3519 replied:
Hmm, taking a step back: what does "is_inplace_on_view" mean in the first place? I can understand what it means for a SavedVariable to be an output, but if "is_inplace_on_view" is True, what does that mean for the SavedVariable?

zou3519 replied:
Does it mean that the SavedVariable holds a Tensor that is the output of the in-place operation on the view?

zou3519 replied:
I read through some more autograd code and convinced myself that this is the case :)

Varal7 replied:
Yes that's exactly correct. An example is 
```
a = torch.randn(6, requires_grad=True)
b = a.clone().view(3, 2)
b.relu_()
```

The relevant call to the `SavedVariable` constructor is https://github.com/pytorch/pytorch/blob/4c3eea26bd80c7126f41ff51d0dcc32f5aae5567/tools/autograd/gen_variable_type.py#L655-L658

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only copied if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary copying data in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever copying `variable` into `data_` if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this copy is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only copied if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

zou3519 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -43,6 +43,18 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+
+    if (get_default_hooks()) {
+      save_metadata(variable);
+      register_hooks_(get_default_hooks(), variable);
```
`register_hooks_` is really confusing to me because it's not obvious what it does. It is shared between this function (the SavedVariable constructor) and `SavedVariable::register_hooks`, but some of the logic only applies for the SavedVariable::register_hooks case. For example, we don't need to do error checking that hooks have already been set and we also don't need to make a check to see if we should save metadata because we're already saving metadata here in the constructor.

Maybe we should rename it to something more descriptive and have it do just the following?
```
set_hooks_and_pack_data(hooks, data) {
    hooks_ = std::move(hooks);
    at::NoGradGuard guard;
    hooks_->call_pack_hook(saved_original_ ? data.tensor_data() : data);
}
```

Varal7 replied:
Makes sense to me, I will adapt the PR.

zou3519 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -90,6 +98,9 @@ void SavedVariable::save_common_metadata(const Variable& data) {
   }
 }
 
+std::unique_ptr<SavedVariableHooks> SavedVariable::get_default_hooks() {
+  return Engine::get_default_engine().get_default_saved_variable_hooks();
+}
```
> I'm using unique_ptr because it makes it easier to guarantee that the underlying PyObject* pack_hook and unpack_hook are correctly increfed and decrefed. This does create a new SavedVariableHooks object with its own pair of hooks for each SavedVariable, but we need that in order for the hooks to survive when default hooks are unset but the SavedVariable is still alive.

Shouldn't the way to model this be std::shared_ptr? In C++, on construction of the `std::shared_ptr<SavedVarableHooks>` we grab the gil and incref the pack_hook and unpack_hook. When the shared_ptr gets deleted, we grab the gil and decref the pack_hook and unpack_hook.

`std::unique_ptr<SavedVariableHooks>` means that every time we construct it, it allocates a SavedVariableHooks on the heap which is something we really want to avoid.

zou3519 replied:
That being said, this is orthogonal to this refactor so it doesn't have to come here in the stack, but I think we should really file an issue and/or fix it

Varal7 replied:
I see your point. I will open an issue about it

Varal7 replied:
@zou3519: After looking into it further with @albanD, it turns out we do need each `SavedVariable` to have its own `PySavedVariableHooks` because that struct stores the `SavedVariable`-specific `PyObject* data_`.
We could overcome this by storing `PyObject* data_` directly into `SavedVariable` but that limits the extensibility of this feature for when we want to implement hooks in C++

zou3519 commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -43,6 +43,18 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+      weak_grad_fn_ = variable.grad_fn();
+    }
+
+    if (get_default_hooks()) {
```
The way this is written right now is not performant. `get_default_hooks()` constructs a unique_ptr, then throws it away, and inside the body we call `get_default_hooks()` again to construct a second unique_ptr.

We should do something like:
```
auto maybe_hooks = get_default_hooks()
if (maybe_hooks) {
  ...
  register_hooks_(std::move(maybe_hooks), variable)
}
```
but really I think we shouldn't be using unique_ptr here (see my later comment)

Varal7 replied:
I will update the PR with this, thanks! And yes I will create open a new PR to think about how to use shared_ptr instead.

zou3519 reviewed (commented):
The refactor looks reasonable, I had a comment about the naming of `register_hooks_` and what it does.

Also had an orthogonal comment about using `std::unique_ptr` for `get_default_hooks()` but that doesn't need to be resolved in this PR

Varal7 committed:
Update on "Avoid unnecessary temporary `data` tensor in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever defining the `data_` tensor if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only needed if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary temporary `data` tensor in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever defining the `data_` tensor if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only needed if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).

Varal7 committed:
Update on "Avoid unnecessary temporary `data` tensor in Saved Variable"


This is a refactor of `SavedVariable.cpp` to prevent ever defining the `data_` tensor if default hooks are set.

Before the refactor:

```c++
data_ = variable.tensor_data(); // this is wasteful if hooks are defined
register_hooks(Engine::get_default_engine().get_default_saved_variable_hooks());
```

After the refactor:
```c++
if (get_default_hooks_()) {
  save_metadata_(variable);
  register_hooks_(get_default_hooks_(), variable);
  return;
}
save_metadata_(variable);
data_ = variable.tensor_data(); // only needed if hooks are not defined
```

Differential Revision: [D29848524](https://our.internmc.facebook.com/intern/diff/D29848524)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848524).
